File: app/src/test/java/com/egecius/rxjava2_demo_2/rx/startWith/StartWithExamplesTest.java
package com.egecius.rxjava2_demo_2.rx.startWith;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;
import java.util.List;

import io.reactivex.observers.TestObserver;

@RunWith(MockitoJUnitRunner.class)
public class StartWithExamplesTest {

    private StartWithExamples mSut;

    @Before
    public void setUp() {
        mSut = new StartWithExamples();
    }

    @Test
    public void startWithAppendsValues() {
        List<Integer> list = Arrays.asList(3, 14);

        TestObserver<Integer> testObserver = mSut.startWith(list).test();

        testObserver
                .assertNoErrors()
                .assertComplete()
                .assertValues(3, 14, 1, 2, 3);
    }

}

File: app/src/test/java/com/egecius/rxjava2_demo_2/rx/startWith/StartWithExamplesTest.kt
package com.egecius.rxjava2_demo_2.rx.startWith

import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.junit.MockitoJUnitRunner
import java.util.*

@RunWith(MockitoJUnitRunner::class)
class StartWithExamplesTest {

    private var mSut: StartWithExamples? = null

    @Before
    fun setUp() {
        mSut = StartWithExamples()
    }

    @Test
    fun startWithObservable() {
        val list = Arrays.asList(3, 14)

        val testObserver = mSut!!.startWithObservable(list).test()

        testObserver
                .assertComplete()
                .assertValues(3, 14, 1, 2, 3)
    }

    @Test
    fun startWithIterable() {
        val list = Arrays.asList(3, 14)

        val testObserver = mSut!!.startWithIterable(list).test()

        testObserver
                .assertComplete()
                .assertValues(3, 14, 1, 2, 3)
    }

    @Test
    fun startWithArray() {
        val array = arrayOf(3, 14)

        val testObserver = mSut!!.startWithArray(*array).test()

        testObserver
                .assertComplete()
                .assertValues(3, 14, 1, 2, 3)
    }

    @Test
    fun startWithElement() {
        val testObserver = mSut!!.startWithElement(14).test()

        testObserver
                .assertComplete()
                .assertValues(14, 1, 2, 3)
    }

}